client/asset/eth/eth.go

```diff
@@ -331,13 +361,32 @@ func (eth *ExchangeWallet) MaxOrder(lotSize uint64, feeSuggestion uint64, nfo *d
 		RealisticWorstCase: realisticWorstCase,
 		RealisticBestCase:  realisticBestCase,
 		Locked:             locked,
-	}, nil
+	}
 }
 
-// PreSwap gets order estimates based on the available funds and the wallet
-// configuration.
-func (*ExchangeWallet) PreSwap(req *asset.PreSwapForm) (*asset.PreSwap, error) {
-	return nil, asset.ErrNotImplemented
+// getInitGas gets an estimate for the gas required for initiating a swap. The
+// result is cached and reused in subsequent requests.
+func (eth *ExchangeWallet) getInitGas() (uint64, error) {
+	eth.cachedInitGasMtx.RLock()
+	cachedInitGas := eth.cachedInitGas
+	eth.cachedInitGasMtx.RUnlock()
+	if cachedInitGas != nil {
+		return *cachedInitGas, nil
+	}
+
+	eth.cachedInitGasMtx.Lock()
+	defer eth.cachedInitGasMtx.Unlock()
```

Comment:
I believe it would be better to hold the write mtx the whole time. I say so because the read would be super quick anyway, and the small gap here allows a space for multiple callers to go for the expensive node calls needlessly, which defeats the benifit of using a mtx over an atomic. imo